### PR TITLE
batch support (prototype)

### PR DESCRIPTION
**Pull request by jonasrauber**
This is a prototype of our upcoming batch support, bringing a massive speed-up to Foolbox without the need to rewrite attacks or even changing the attack logic to support batches (which can be very difficult for certain attacks and comes with other disadvantages).

So far, this works with `TensorFlowModel` and `PyTorchModel`. You can try this now using `CarliniWagnerAttack`, `GradientAttack` and similar ones and all `PGD`-like attacks (`BasicIterativeMethod`, `RandomStartProjectedGradientDescentAttack`, etc.).

The current API to try this feature looks like this:

```python3
from foolbox.batching import run_parallel_attack

model = ...
images = ...
labels = ...

attack_create_fn = foolbox.attacks.CarliniWagnerL2Attack  # for example
criterion = foolbox.criteria.Misclassification()  # for example

advs = run_parallel_attack(attack_create_fn, model, criterion, images, labels)
# advs will be a list of Adversarial objects
```

This Jupyter notebook contains a complete example: https://gist.github.com/jonasrauber/140ada5a352cabb3dd0e91b9cd3adf03

### CODE REVIEW

**Comment by jonasrauber** [comment]
@wielandbrendel please give it a try

**Comment by jonasrauber** [comment]
This PR will solve #208

**Comment by jvrsgsty** [comment]
Giving this feature a try, I realized that it relies on the `yield from` expression which was introduced in Python 3.3. I know Python 2.7 EOL is coming up. Is there no intent to make the code backwards compatible with Python 2.7?

**Comment by jonasrauber** [comment]
No, new features will no longer support Python 2.7 (also true for NumPy, btw) and we might even start removing Python 2 support explicitly to simplify the code.

**Comment by yfreedomliTHU** [comment]
Which version is required to use run_parallel_attack?I attempted it on version 1.9.0, but  it raised an error: **"No module named 'foolbox.batching'"**

**Comment by jonasrauber** [comment]
It will be part of Foolbox 2.0.
To try it out now, you need to pull this PR and install it.

**Comment by yfreedomliTHU** [comment]
@jonasrauber Ok, I got it! Thank you!

**Comment by yfreedomliTHU** [comment]
@jonasrauber,I have installed the PR with 'batching.py' , I trained my own tensorflow model and I transfered my tf model to fmodel according to the document.But the error  " **Can not squeeze dim[0], expected a dimension of 1, got 22 for 'Squeeze_1' (op: 'Squeeze') with input shapes: [22,110].**" still exists. It seems that I still can't attack my own tf model with batch input.
Thank you!

**Comment by jonasrauber** [comment]
Please have a look at the Jupyter notebook I linked above that contains a full example and compare shapes.

**Comment by yfreedomliTHU** [comment]
> Please have a look at the Jupyter notebook I linked above that contains a full example and compare shapes.

I have read it , but the model in the Jupyter notebook was downloaded from utrl by zoo.get_model(). And I want to attack my own model(trained with tensorflow), which means I have to transfer my model to fmodel according to the document. However, when I transfer my own tf_model to fmodel with the code:**fmodel = foolbox.models.TensorFlowModel(images, logits, (0,1))**,it raise the error:" **Can not squeeze dim[0], expected a dimension of 1, got 22 for 'Squeeze_1' (op: 'Squeeze') with input shapes: [22,110]**." I have already installed the PR, and it seems that I still can't transfer my own model to fmodel
 with batch imput.
Thank you very much for your answer！

**Comment by jonasrauber** [comment]
Probably this has nothing to do with this PR. Creating the fmodel is the same with and without it, and the tensors should always have a batch dimension. Please open a new issue with more details on your model.

**Comment by jonasrauber** [comment]
superseded by #316

**Comment by jonasrauber** [comment]
Finally released as Foolbox 2.0.0b0.
